packages/connect/src/ScProvider/Health.ts

```diff
@@ -0,0 +1,319 @@
+      // request.
+      this.startHealthCheck()
+      this.update()
+      return null
+    }
+
+    // Response doesn't concern us.
+    if (parsedResponse.id) {
+      // Need to remove the `extern:` prefix.
+      if (!parsedResponse.id.startsWith("extern:"))
+        throw new Error("State inconsistency in health checker")
+      const newId = JSON.parse(
+        parsedResponse.id.slice("extern:".length),
+      ) as string
+      parsedResponse.id = newId
+    }
+
+    return JSON.stringify(parsedResponse)
+  }
+
+  update = (): void => {
+    if (!this.#currentHealthTimeout) {
+      this.#currentHealthTimeout = setTimeout(() => {
+        this.#currentHealthTimeout = null
+        this.startHealthCheck()
+      }, 10000)
+    }
+
+    if (
+      this.#isSyncing &&
+      !this.#currentSubscriptionId &&
+      !this.#currentSubunsubRequestId
+    )
+      this.startSubscription()
+    if (
+      !this.#isSyncing &&
+      this.#currentSubscriptionId &&
+      !this.#currentSubunsubRequestId
+    )
+      this.endSubscription()
+  }
+
+  startHealthCheck = (): void => {
+    if (this.#currentHealthCheckId)
+      throw new Error("Internal error in health checker")
+    if (this.#currentHealthTimeout) {
+      clearTimeout(this.#currentHealthTimeout)
+      this.#currentHealthTimeout = null
+    }
+    this.#currentHealthCheckId = "health-checker:".concat(
+      this.#nextRequestId.toString(),
+    )
+    this.#nextRequestId += 1
+    this.#requestToSmoldot(
+      JSON.stringify({
+        jsonrpc: "2.0",
+        id: this.#currentHealthCheckId,
+        method: "system_health",
+        params: [],
+      }),
+    )
+  }
+
+  startSubscription = (): void => {
+    if (this.#currentSubunsubRequestId || this.#currentSubscriptionId)
+      throw new Error("Internal error in health checker")
+    this.#currentSubunsubRequestId = "health-checker:".concat(
+      this.#nextRequestId.toString(),
+    )
+    this.#nextRequestId += 1
+    this.#requestToSmoldot(
+      JSON.stringify({
+        jsonrpc: "2.0",
+        id: this.#currentSubunsubRequestId,
+        method: "chain_subscribeNewHeads",
+        params: [],
+      }),
+    )
+  }
+
+  endSubscription = (): void => {
+    if (this.#currentSubunsubRequestId || !this.#currentSubscriptionId)
+      throw new Error("Internal error in health checker")
+    this.#currentSubunsubRequestId = "health-checker:".concat(
+      this.#nextRequestId.toString(),
+    )
+    this.#nextRequestId += 1
+    this.#requestToSmoldot(
+      JSON.stringify({
+        jsonrpc: "2.0",
+        id: this.#currentSubunsubRequestId,
+        method: "chain_unsubscribeNewHeads",
+        params: [this.#currentSubscriptionId],
+      }),
+    )
+  }
+
+  destroy = (): void => {
+    if (this.#currentHealthTimeout) {
+      clearTimeout(this.#currentHealthTimeout)
+      this.#currentHealthTimeout = null
+    }
+  }
+}
+
+export class HealthCheckError extends Error {
```

Comment:
I would suggest to completely remove this error (either in this PR or another), because there's no point in gracefully handling bugs in smoldot.

Reply:
Will address in different PR - if thats ok with you